### ISSUE TITLE
Orders Map Data Source

### ISSUE THREAD

**Issue by hverdonk** [vsoch/freegenes#112]
Where does the orders map pull location data from? I know that not all of the Shipments have tracking numbers/locations associated with them, but it seems like the map may not be updating when new shipments are entered, and I'd like to be sure before I submit a bug report.

**Comment by vsoch** [comment]
It's generating the data file on a nightly basis from the Shippo API. The only thing represented are zip codes, and it's sized based on the count. I just ran it fresh and compared to the current data file, they are identical:

```
{
    "60177": {
        "lat": 41.99,
        "long": -88.31,
        "count": 2
    },
    "11217": {
        "lat": 40.68,
        "long": -73.98,
        "count": 1
    },
    "60660": {
        "lat": 41.98,
        "long": -87.66,
        "count": 1
    },
    "V6A 1M3": {
        "lat": 49.2818536,
        "long": -123.0876828,
        "count": 1
    }
}
```
So unfortunately if all the shipments are not represented, it's likely another rate limit that is being set by the Shippo API (akin to how we only got 3 rates for a shipment). I'm taking a look now.

**Comment by vsoch** [comment]
ha, so when the paginated response returns, it provide a URL for the next page, that looks like this:

```
https://api.goshippo.com/orders/?page=2
```
And guess what? When we actually ask for that page, we get 404. Way to go Shippo, provide a URL that doesn't work.

**Comment by vsoch** [comment]
And note that we're following their docs to a T https://goshippo.com/docs/orders/#list-orders

**Comment by hverdonk** [comment]
Ah, of course. Well, all I can do is give you a little more information on what you SHOULD be seeing to help with debugging/talking to Shippo. Currently we have 7 shipments total in Shippo itself (which doesn't match up to what's on our website, because some were made through FedEx, yadda yadda). Shippo is returning the 5 most recent shipments, so presumably the results are ordered most to least recent.

**Comment by vsoch** [comment]
We'll need to open up an issue for this, and ask them what GET param can be used to (correctly) paginate, OR ask why the next URL is 404. I am looking in their docs and I don't see anything that sets a page or limit (but I do see Orders is flagged with "beta" so we are probably the first to see this).

**Comment by vsoch** [comment]
Here is what you can say:

Hi Shippo,

We are using your beta Orders endpoint to retrieve our orders, and we were trying to figure out why only 5 were returned (there should be more). We investigated the logic of your curl endpoints, and although the first response (to https://api.goshippo.com/orders/) returns 200 with keys "results" (a small list of orders) and then "next":

results["next"]
https://api.goshippo.com/orders/?page=2

When we do the same call to this URL, a 404 is returned, and this shouldn't be the case.

response = requests.get(results["next"], headers=headers) 
 
The response is 404 (not found)

response                                                                                                                    
<Response [404]>

Could you please tell us how to properly paginate your orders API endpoint?
We are following the instructions in your docs, and can't find any detail about
GET parameters for setting a larger limit or asking for a specific page. Thanks!

Best,

(and feel free to CC me).

**Comment by hverdonk** [comment]
I'll send off a new request to them, copying what you've written here. The initial support request is made directly through their website and doesn't seem to support CC-ing, so I'll just forward you the response I receive.

**Comment by vsoch** [comment]
Sounds good.

**Comment by hverdonk** [comment]
Alright, here's the response I got:

> Thank you for reaching out to us about issues retrieving orders. I apologize for the delay in us getting back to you. We are still recovering from an influx of support requests. We truly appreciate your patience!
> 
> After clicking the link to page 2 the rest of the orders should return in the response after entering your headers. I tested this with own account in Postman and did not have trouble retrieving orders on the additional pages. I want to dive into this deeper for you because I see there are seven orders in the account under drew_endy_lab@stanford.edu so you should be able to retrieve the additional two orders from page 2. I tried to test retrieving orders with your account but, the account already has two tokens generated so I couldn't. Can you send me the API request and response in JSON in a text file so I can investigate the 404 error further?

**Comment by vsoch** [comment]
Yep, give me a few minutes to reproduce this for them! It's funny that they can't test our account with the tokens (actually maybe that's a security thing and it should be like that) but either way I'm really happy they are going to look into this. Give me a few minutes!

**Comment by vsoch** [comment]
okay all set! It's a private gist so I'll send this to you in an email.

**Comment by hverdonk** [comment]
Thanks! I've sent them the link to the gist page.

**Comment by hverdonk** [comment]
This is what I wrote to Shippo tech support:

> Thanks for getting back to me. This link<https://gist.github.com/vsoch/9670c28196510acf07632811dcd29553> has a whole run-through of how to reproduce the error, including the API request and the response JSON.

And this was the response:

> You are welcome for the reply.
> 
> Thanks for the GitHub link. Unfortunately, that does not help me with sorting out the issue. However, our engineering and developer outreach teams will review the repo and assist you.
> 
> You can also use StackOverflow for more general help from other developers who have implemented Shippo before, as well as from our developer outreach team. Just be sure to use the tag “shippo” so our outreach team can see it and assist you.

**Comment by vsoch** [comment]
Seriously? That's terrible customer service - we've given them exact code (and output) to reproduce the error, I'm not sure what to do beyond that. Directing us to stack overflow is... insulting.

**Comment by vsoch** [comment]
This seems like our only hope?

> However, our engineering and developer outreach teams will review the repo and assist you.

**Comment by vsoch** [comment]
Maybe we should ask them what _would_ help with sorting out the issue?

**Comment by vsoch** [comment]
At least it's comforting that others on Stack Overflow are hitting some of the same issue we are:

 - [here is not getting rates returned](https://stackoverflow.com/questions/48735789/rates-not-returning-while-creating-a-shipment-using-shippo)

There are also only [52 questions](https://stackoverflow.com/questions/tagged/shippo?tab=newest&page=1&pagesize=15), many without answers, so I'm wondering what questions shippo has in mind for us to read, or if the suggestion is to post, who they expect to answer? 

TLDR, their response is hugely a cop out. In that this is a paid service, I would see if it's possible to go a bit higher up in their support queue, maybe to a manager? It's unacceptable for a business to operate like this, imho.

**Comment by hverdonk** [comment]
> we've given them exact code (and output) to reproduce the error  

Not only did we give them what they needed to fix the issue, it was _exactly what they asked for in the first place._ It's like she didn't even look at my response once she didn't see a text file.

>Directing us to stack overflow is... insulting.

I completely agree.

>Maybe we should ask them what *would* help with sorting out the issue?

I suspect the problem is on their side, but they assume that it's our code that has the bug and aren't willing to examine our code (which, fair, but at least examine your **own** code). We could still ask as a way to motivate them to look into the issue (I agree, the engineering and developer outreach teams are our only hope at this point), but if they aren't even going to look at the content they ask us to provide, I'm not sure where to go from here. 

@Koeng101, we're gonna have to talk to Marc about this one.

**Comment by vsoch** [comment]
Minimally they could send us a request, heck in any language (curl likely is the most basic) that they can verify is working. They keep citing [postman](https://www.getpostman.com/) which is a wrapper / interface for working with APIs, it doesn't do much other than make it easier to do these same requests, or bulk requests (akin to using Git command line vs. some Windows app for it). So, they should be able to copy a fully working request from their postman to share with us... it's a totally reasonable thing for a business to provide a working example for an endpoint.

**Comment by vsoch** [comment]
hey! So I poked around the Shippo API again and I found that (some) shipments carry order ids! So i tried deriving a unique set from the shipments, and then using the Shippo Orders endpoint to retrieve order metadata one at a time. It's a really inefficient query compared to how it should work, but since it's a nightly job the slight slowness should be okay! And I think we are good, now the map is updated with all the orders -> https://freegenes.dev/map/orders and see the changes here https://github.com/vsoch/freegenes/commit/2af56d7b93bc6cead1acc88e5c8f0aa06c809608. If this looks okay, then I'd still voice concern to Shippo (re: discussion above) but minimally I think this should close the issue.

You know how they have those "Bad dog warning signs?" I'd say a "Bad API warning sign" is when you have to use other endpoints as roundabout ways to get the content you actually need, because the main endpoints are broken.

**Comment by vsoch** [comment]
What do you think about changing the zoom and center for the map? I just noticed there is an order in Cambridge, and it would be super cool to see it.

**Comment by vsoch** [comment]
okay I moved it over slightly, zoomed out, to add Europe

![image](https://user-images.githubusercontent.com/814322/68327473-6b660380-009b-11ea-98b9-7c5dd279c0ea.png)

**Comment by hverdonk** [comment]
It looks great! This should work for now. In the future, when we start storing lab address data on the server, we'll just use those addresses to populate the map, negating the need for Shippo's broken API endpoint.
